opennlp-tools/src/main/java/opennlp/tools/ml/model/ModelParameterChunker.java

```diff
@@ -0,0 +1,142 @@
+ * directly followed by the number of chunks used. This way, when reading in chunked model parameters,
+ * recombination is achieved transparently.</li>
+ * </ul>
+ * <p>
+ * Note: Both, existing (binary) model files and newly trained models which don't require the chunking
+ * technique, will be supported like in previous OpenNLP versions.
+ *
+ * @author <a href="mailto:martin.wiesner@hs-heilbronn.de">Martin Wiesner</a>
+ * @author <a href="mailto:struberg@apache.org">Mark Struberg</a>
+ */
+public final class ModelParameterChunker {
+
+  /*
+   * A signature that denotes the start of a String that required chunking.
+   *
+   * Semantics:
+   * If a model parameter (String) carries the below signature at the very beginning, this indicates
+   * that 'n > 1' chunks must be processed to obtain the whole model parameters. Otherwise, those would not be
+   * written to the binary model files (as reported in OPENNLP-1366) if the training occurs on large corpora
+   * as used, for instance, in the context of (very large) German NLP models.
+   */
+  public static final String SIGNATURE_CHUNKED_PARAMS = "CHUNKED-MODEL-PARAMS:"; // followed by no of chunks!
+
+  private static final int MAX_CHUNK_SIZE_BYTES = 65535; // the maximum 'utflen' DataOutputStream can handle
+
+  private ModelParameterChunker(){
+    // private utility class ct s
+  }
+
+  /**
+   * Reads model parameters from {@code dis}. In case the stream start with {@link #SIGNATURE_CHUNKED_PARAMS},
+   * the number of chunks is detected and the original large parameter string is reconstructed from several
+   * chunks.
+   *
+   * @param dis   The stream which will be used to read the model parameter from.
+   */
+  public static String readUTF(DataInputStream dis) throws IOException {
+    String data = dis.readUTF();
+    if (data.startsWith(SIGNATURE_CHUNKED_PARAMS)) {
+      String chunkElements = data.replace(SIGNATURE_CHUNKED_PARAMS, "");
+      int chunkSize = Integer.parseInt(chunkElements);
+      StringBuilder sb = new StringBuilder();
+      for (int i = 0; i < chunkSize; i++) {
+        sb.append(dis.readUTF());
+      }
+      return sb.toString(); // the reconstructed model parameter string
+    } else {  // default case: no chunked data -> just return the read data / parameter information
+      return data;
+    }
+  }
+
+  /**
+   * Writes the model parameter {@code s} to {@code dos}. In case {@code s} does exceed
+   * {@link #MAX_CHUNK_SIZE_BYTES} in length, the chunking mechanism is used; otherwise the parameter is
+   * written 'as is'.
+   *
+   * @param dos   The {@link DataOutputStream} stream which will be used to persist the model.
+   * @param s     The input string that is checked for length and chunked if {@link #MAX_CHUNK_SIZE_BYTES} is
+   *              exceeded.
+   */
+  public static void writeUTF(DataOutputStream dos, String s) throws IOException {
+    try {
+      dos.writeUTF(s);
+    } catch (UTFDataFormatException dfe) {
```

Comment:
Is it possible to check the length of `s` instead? I think what I'm asking is can this be done by some logic instead of in a try/catch.

Reply:
@jzonthemtn In general, this looks entirely plausible. Yet, there are specific details for encoding characters in UTF(8) format. Therefore, byte length of `s` is not sufficient or valid for all cases. 

Thus, @struberg and I decided to leave to the actual implementation of `DataOutputStream.writeUTF(...)` to tell us the truth. Have a look and check the code there, if interested, for the "difficult" part. We decided not to cnp the code over to OpenNLP, but to rely on the checks made in `DataOutputStream`. 

I hope, you can follow our idea/decision to go for "avoid cnp" in favor of catching `UTFDataFormatException` here once, in case the limit exceeds 64k. This seems acceptable for those "rare" occasions when training with large corpora. 

As a background: The 'Tueba Wikipedia' treebank only runs into this exception twice; other (well- known) German corpora, like the one from Hamburg I also used for validation, hit it 3 or 4 times during training.

Reply:
I figured you had a good reason for it. I think it's fine.